snakemake/executors/flux.py

```diff
@@ -0,0 +1,266 @@
+        if not flux:
+            sys.exit(
+                "Cannot import flux. Is a cluster available to you with Python bindings?"
+            )
+        self._fexecutor = flux.job.FluxExecutor()
+
+    def cancel(self):
+        """
+        cancel execution, usually by way of control+c. Cleanup is done in
+        shutdown (deleting cached workdirs in Google Cloud Storage
+        """
+        for job in self.active_jobs:
+            if not job.flux_future.done():
+                flux.job.cancel(self.f, job.jobid)
+        self.shutdown()
+
+    def _set_job_resources(self, job):
+        """
+        Given a particular job, generate the resources that it needs,
+        including default regions and the virtual machine configuration
+        """
+        mem_mb = job.resources.get("mem_mb", 15360)
+
+        # Update default resources
+        self.default_resources = DefaultResources(
+            from_other=self.workflow.default_resources
+        )
+        self.default_resources.set_resource("mem_mb", mem_mb)
+
+    def get_snakefile(self):
+        assert os.path.exists(self.workflow.main_snakefile)
+        return self.workflow.main_snakefile.replace(self.workdir, "").strip(os.sep)
+
+    def _get_jobname(self, job):
+        # Use a dummy job name (human readable and also namespaced)
+        return "snakejob-%s-%s-%s" % (self.run_namespace, job.name, job.jobid)
+
+    def _generate_environment(self):
+        """
+        Loop through envvars (keys to host environment) and add
+        any that are requested for the runtime environment
+        """
+        envvars = dict(os.environ)
+        for key in self.envvars:
+            try:
+                envvars[key] = os.environ[key]
+            except Exception:
+                continue
+        return envvars
+
+    def format_job_exec(self, job):
+        """
+        We don't want a python -m to run.
+        """
+        prefix = self.get_job_exec_prefix(job)
+        if prefix:
+            prefix += " &&"
+        suffix = self.get_job_exec_suffix(job)
+        if suffix:
+            suffix = f"&& {suffix}"
+        return join_cli_args(
+            [
+                prefix,
+                self.get_envvar_declarations(),
+                "snakemake",
+                format_cli_arg("--snakefile", self.get_snakefile()),
+                self.get_job_args(job),
+                self.general_args,
+                suffix,
+            ]
+        )
+
+    def run(self, job, callback=None, submit_callback=None, error_callback=None):
+        """
+        Submit a job to flux.
+        """
+        super()._run(job)
+
+        from flux.job import JobspecV1
+
+        # This needs to be called after the line above
+        self._prepare_job_formatter()
+
+        # Prepare job resourcces
+        self._set_job_resources(job)
+
+        # The entire snakemake command to run, etc
+        command = self.format_job_exec(job)
+        nodes = job.resources.get("_nodes")
```

Comment:
MPI?

Reply:
I'm actually not sure! I'm not familiar enough with flux to be able to answer, but I'm happy to start with one node and we can tweak if someone opens an issue.

Reply:
yep confirmed we can do MPI! So okay to leave as is?

Reply:
I am still seeking for a proper way to reflect MPI in Snakemake rules. This will then however most likely not be via nodes resource, but rather some mpi specific keyword. Hence, I'd not make use of nodes here and keep this to the default in flux, which is likely 1, correct?

Reply:
gotcha!